google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITForeignKeyDeleteCascadeTest.java

```diff
@@ -0,0 +1,540 @@
+        // Use 'p1' to bind to the parameter with index 1 etc.
+        .bind("p1").to(3L)
+        .bind("p2").to("singerName").build();
+
+    final String concertInsertStatement = getInsertStatementForConcertsTable();
+    final Statement concertInsertStatementWithValues = Statement.newBuilder(concertInsertStatement)
+        // Use 'p1' to bind to the parameter with index 1 etc.
+        .bind("p1").to(3L)
+        .bind("p2").to(3L).build();
+
+    // successful inserts into referenced and referencing tables
+    databaseClient
+        .readWriteTransaction()
+        .run(
+            transaction -> {
+              transaction.executeUpdate(singerInsertStatementWithValues);
+              return null;
+            });
+    databaseClient
+        .readWriteTransaction()
+        .run(
+            transaction -> {
+              transaction.executeUpdate(concertInsertStatementWithValues);
+              return null;
+            });
+
+    // execute delete
+    final String singerDeleteStatement = getDeleteStatementForSingerTable();
+    final Statement singerDeleteStatementWithValues = Statement.newBuilder(singerDeleteStatement)
+        // Use 'p1' to bind to the parameter with index 1 etc.
+        .bind("p1").to(3L).build();
+    databaseClient
+        .readWriteTransaction()
+        .run(
+            transaction -> {
+              transaction.executeUpdate(singerDeleteStatementWithValues);
+              return null;
+            });
+
+    try (ResultSet resultSet =
+        databaseClient.singleUse()
+            .executeQuery(Statement.of("SELECT * FROM " + TABLE_NAME_SINGER))) {
+      assertThat(resultSet.next()).isFalse();
+    }
+
+    try (ResultSet resultSet =
+        databaseClient.singleUse()
+            .executeQuery(Statement.of("SELECT * FROM " + TABLE_NAME_CONCERTS))) {
+      assertThat(resultSet.next()).isFalse();
+    }
+  }
+
+  @Test
+  public void testForeignKeyDeleteCascadeConstraints_forMutations_onConflictDueToParentTable() {
+    final DatabaseClient databaseClient = getCreatedDatabaseClient();
+
+    // inserting and deleting the referenced key within the same mutation are considered
+    // conflicting operations, thus this results in an exception.
+    try {
+      databaseClient
+          .readWriteTransaction()
+          .run(
+              transaction -> {
+                transaction.buffer(
+                    Arrays.asList(
+                        Mutation.newInsertBuilder("Singer")
+                            .set(getSingerIdColumnName()).to(4L)
+                            .set(getSingerFirstNameColumnName()).to("singerName").build(),
+                        Mutation.delete("Singer", Key.of(4L))));
+                return null;
+              });
+      fail("Expected exception");
+    } catch (SpannerException ex) {
+      assertThat(ex.getErrorCode()).isEqualTo(ErrorCode.FAILED_PRECONDITION);
+    }
```

Comment:
Yes, I had this same query since it looked unrelated to this feature but we got a guidance from backend team to include it in the set of tests. @asthamohta Are we including this (#7 test case in spec document) in other libraries?